### DISCUSSION TITLE
Tensorial operators, tensor conversion, and helper functions

### DISCUSSION THREAD

**Discussion by YangZhao16** [thelfer/tfel#121] [Q&A]
Hi,

I was working on a mfront file trying to customize a material behavior, and I ran into the following issues:

1. Is there a defined operator to get the derivative of ({a}\dot{b}\dot{c}) with respective to {b}, where {a},{b},{c} are all rank-2 symmetric tensor? I find a function 'symmetric_product_derivative_daba_db', but it requires {a} = {c}.
In fact, I want to get the rank-4 tensor A such that A_ijkl = a_ik * b_jl + a_il * b_jk.

2. Once we construct a rank-4 tensor in 3d space, is there any function that we can call to reduce it to 2d space under plane strain assumption? I am asking this because I can manually assemble a rank-4 tensor in 3D for my material, but I still need to reduce it to plane strain assumption somehow.

3. Within the front file, is there a way to define a helper function (e.g., the tensorial operator I mentioned in 1), or do we have to define it elsewhere?

Best,
Yang.

**Comment by thelfer** [comment]
@YangZhao16 Thanks for your interest to `MFront`.
Great to see that you already read the documentation in depth. I'll (try to) answer your questions one by one.

**Comment by thelfer** [comment]
Regarding Question 1, the trouble is that `a*b*c` is a non symmetric tensors, so the derivative with respect to b is an object of type `st2tot2` (i.e. a linear transformation from symmetric tensor to a non symmetric tensor). There is no standard operations to do this in `TFEL`.

If we convert those tensors to non symmetric tensors, then we could compute the derivatives by chain rule using the `t2tot::trld` (tensor product left derivative) , but the results will be a `to2to2` which is probably not want you want... However,  we could take that `to2to2`  and multiply it by the `st2tot2` which converts a symmetric tensor to a non symmetric tensors.

So we have two choices here:

1. Give me some insights on where does this operation comes from and may be find a workaround.
2. Implement a function returning the `st2tot2` which converts a symmetric tensor to a non symmetric tensors and use chain rule. Fairly straighforward and probably a missing piece in `TFEL/Math`
3. Implement a function which, given `a`, `b` and `c`, computes the derivative that you want. A bit combersome but feasible.

**Reply by YangZhao16** [reply]
Hi,

Basically, I was trying to implement transversely isotropic elasticity with a tensorial formulation introduced by Spencer in the book "Constitutive Theory for Strongly Anisotropic Solids", 1984, page 6, to avoid using a tangent matrix assuming the bedding structure is aligned with the coordinate system and rotating the tangent matrix and the stress term to the global coordinate system. Maybe it's better off using the later scheme, since there already exists well-developed code in Mfront.

![123](https://user-images.githubusercontent.com/29029186/154839873-c0a2d8eb-45ee-4874-9645-7b9e9111e6d9.PNG)

**Reply by thelfer** [reply]
I am not sure that have understand your point. You want to avoid the rotation in the material frame ?
Because otherwise, defining an isotropic elastic transverse stiffness matrix is quite straightforward using the `@ComputeStiffnessTensor`.

**Reply by YangZhao16** [reply]
Hi @thelfer, 

In my case, the material frame is not aligned with the global frame (i.e., inclined bedding planes), so if I define a stiffness matrix in the material frame, I need to somehow convert it to the global frame, right?

I find a discussion on implementing transversely isotropic elasticity in mfront, see
https://sourceforge.net/p/tfel/discussion/mechanicalbeahviours/thread/ca070657cb/
I wonder in this case, how can we define the stiffness matrix with just 4 parameters? Since for transversely isotropic elastic material, there are in fact 5 independent parameters in the tangent matrix.

**Reply by thelfer** [reply]
> I need to somehow convert it to the global frame, right?

No, you don't have to do anything.

`MFront` handles automatically the rotation from the global frame to the material frame on input (for strain and deformation gradient) and the rotation from the material frame to the global frame for the (for the stress and consistent tangent matrix). 

Note that the internal state variables are always assumed to be in the material frame.

The way this rotation is done depends on the interface considered. Some codes already handle those rotations, so `MFront` does nothing in this case.

> I wonder in this case, how can we define the stiffness matrix with just 4 parameters? Since for transversely isotropic elastic material, there are in fact 5 independent parameters in the tangent matrix.

In this example, we used the `@ComputeStiffnessTensor` keyword, which expects 9 parameters in general (3 Young' moduli, 3 Poisson' ratios, 3 shear moduli).  I **never** really has to treat  transversely isotropic elastic material, so I never though of how to reduce this list of coefficients.

See:

~~~~
$ mfront --help-keyword=Implicit:@ComputeStiffnessTensor
~~~~

The output of which is:

The `ComputeStiffnessTensor` keyword is used to define the elastic stiffness tensor based on the elastic material properites given as an array of entries. After this array, a semi-colon is expected.

This array is used to automatically used to declared the elastic material properties of the behaviour (see the `@ElasticMaterialProperties` keyword for details).

An entry can be either a string referring to an external `MFront` file, a formula or a numerical value.

If an entry refers to an external `MFront` file or the formula, all the inputs of this material property must be either:

- a material property
- a parameter
- a state variable
- an external state variable

of the behaviour.

## Isotropic case

In the isotropic case, two entries are expected in the array, in that order:

- the Young Modulus
- the Poisson ratio

## Orthotropic case

In the orthotropic case, 9 entries are expected in the array, in that order:

- three Young modulus \((E_{1},E_{2},E_{3})\)
- three Poisson ratio \((nu_{12},nu_{23},nu_{13})\)
- three shear modulus \((G_{12},G_{23},G_{13})\)

In the orthoropic case, computation of the stiffness tensor rely on the definition of an orthotropic convention. For example, the `Pipe` orthotropic convention will lead to automatically exchange the second and first axes when computing the stiffness tensor for the plane strain, plane stress and generalised plane strain hypotheses. See the `OrthotropicBehaviour` keyword for details.

## Tensors computed

In an implicit scheme, this keyword leads to the definition and the automatic computation of the tensors `D` and `D_tdt` which respectively refer to the stiffness tensor at \(t+\theta\,dt\) and \(t+dt\).

In an explicit scheme, this this keyword leads to the definition of the stiffness tensor which is automatically updated if this tensor evolves during the time step.

## Example

~~~~ {#ComputeStiffnessTensor .cpp}
@ComputeStiffnessTensor{7.8e+4,2.64233e+5,3.32e+5,
    0.13,0.24,0.18,
    4.8e+4,1.16418e+5,7.8e+4};
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Reply by YangZhao16** [reply]
Hi @thelfer,

Thank you very much for your quick and detailed reply!

From the conventions you mentioned, I now know how to reduce orthotropic materials to transversely isotropic materials.

Just one quick question, how do we specify the spatial relations between material frame against global frame, so that the code knows that I am investigating a problem with an inclined bedding structure?

**Reply by thelfer** [reply]
> Just one quick question, how do we specify the spatial relations between material frame against global frame, so that the code knows that I am investigating a problem with an inclined bedding structure?

I depends on the code considered ! You must look in the documentation of the code that you want to use how to define the rotation matrix.

**Reply by YangZhao16** [reply]
Got it, thanks! I will go through the documentation thoroughly. Thank you for your assistance!

**Comment by thelfer** [comment]
Regarding Question 2, no this feature does not exists. Note that the tensorial operations in `TFEL/Math` only depends on the space dimension, not the modelling hypothesis.

**Reply by thelfer** [reply]
This means that probably need to implement it by hand. If you do it, do not hesitate to contribute to `TFEL/Math` (open an issue about it).

**Comment by thelfer** [comment]
Regarding Question 3, you have several possibilities depending on your taste, but the basis is to use the `@Includes` keyword as follows:

```cxx
@Includes{
#ifndef YANG_ZHAO_TFEL_MATH_EXTENSIONS
#ifndef YANG_ZHAO_TFEL_MATH_EXTENSIONS 1

namespace tfel::math{

<your functions here>

}

#endif
}
```

You can also use an external header file as follows:

```cxx
@Includes{
#include "MyExtensions.hxx"
}
```

You can then use the `-I` option to tell `MFront` where your headers are. Generally, there are in the current directory, so this works fine:

```bash
$ mfront -I $(pwd) ....
```

**Reply by YangZhao16** [reply]
Thank you very much for your detailed reply! I really appreciate your support and guidance! @thelfer
